src/librbd/io/Utils.cc

```diff
@@ -121,6 +122,58 @@ void read_parent(I *image_ctx, uint64_t object_no, const Extents &extents,
                             ReadResult{data}, 0, trace);
 }
 
+uint64_t get_extents_total_length(Extents &extents) {
+  uint64_t total_bytes = 0;
+  for (auto& image_extent : extents) {
+    total_bytes += image_extent.second;
+  }
+  return total_bytes;
+}
+
+template <typename I>
+int clip_request(I *image_ctx, Extents &image_extents) {
+  std::shared_lock image_locker{image_ctx->image_lock};
+  for (auto &image_extent : image_extents) {
+    auto clip_len = image_extent.second;
+    int r = clip_io(librbd::util::get_image_ctx(image_ctx), image_extent.first, &clip_len);
+    if (r < 0) {
+      return r;
+    }
+
+    image_extent.second = clip_len;
+  }
+  return 0;
+}
+
+void set_read_clip_length(Extents &image_extents,
+                          AioCompletion *aio_comp) {
+  uint64_t buffer_length = 0;
+  for (auto &image_extent : image_extents) {
+    buffer_length += image_extent.second;
+  }
+  aio_comp->read_result.set_clip_length(buffer_length);
+}
+
+template <typename I>
+bool finish_request_early_if_readonly(I *image_ctx, AioCompletion *aio_comp) {
```

Comment:
Nit: this methods will probably move, but just in case:
```suggestion
bool finish_request_early_if_read_only(I *image_ctx, AioCompletion *aio_comp) {
```

Reply:
Same as above.